File: ExercismFSharp/ExercismFSharp/SecretHandshake.fs
module SecretHandshake

open System

type private HandshakeMovement =
    | Wink              = 1
    | DoubleBlink       = 2
    | CloseYourEyes     = 4
    | Jump              = 8
    | ReverseOperations = 16

let handshake num =
    let isIncludedMovement mov =
        num &&& int mov <> 0

    let stringMovement mov =
        match mov with
        | HandshakeMovement.Wink          -> "wink"
        | HandshakeMovement.DoubleBlink   -> "double blink"
        | HandshakeMovement.CloseYourEyes -> "close your eyes"
        | HandshakeMovement.Jump          -> "jump"
        | _                               -> failwith "no string representation for reverse"

    let movements =
        Enum.GetValues(typeof<HandshakeMovement>)
        |> Seq.cast<HandshakeMovement>
        |> Seq.filter (fun x -> x <> HandshakeMovement.ReverseOperations)
        |> Seq.filter isIncludedMovement
        |> Seq.map stringMovement
        |> Seq.toList

    if isIncludedMovement HandshakeMovement.ReverseOperations then
        movements |> List.rev
    else
        movements

File: Exercism/python/secret-handshake/secret_handshake.py
def commands(number):
    codes = [
        (1, "wink"),
        (2, "double blink"),
        (4, "close your eyes"),
        (8, "jump"),
    ]

    result = []

    for value, action in codes:
        if number & value != 0:
            result.append(action)

    if number & 16 != 0:
        result.reverse()

    return result


def secret_code(actions):
    codes = {
        "wink": 1,
        "double blink": 2,
        "close your eyes": 4,
        "jump": 8
    }

    result = 0

    for action in actions:
        result |= codes[action]

    reverse = False
    for i in range(0, len(actions) - 1):
        if codes[actions[i]] > codes[actions[i + 1]]:
            reverse = True
            break

    if reverse:
        result |= 16

    return result
